crates/ide-assists/src/handlers/convert_two_arm_bool_match_to_matches_macro.rs

```diff
@@ -0,0 +1,293 @@
+use syntax::ast::{self, AstNode, Pat};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: convert_two_arm_bool_match_to_matches_macro
+//
+// Convert 2-arm match that evaluates to a boolean into the equivalent matches! invocation.
+//
+// ```
+// fn main() {
+//     match scrutinee$0 {
+//         Some(val) if val.cond() => true,
+//         _ => false,
+//     }
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     matches!(scrutinee, Some(val) if val.cond())
+// }
+// ```
+pub(crate) fn convert_two_arm_bool_match_to_matches_macro(
+    acc: &mut Assists,
+    ctx: &AssistContext<'_>,
+) -> Option<()> {
+    let match_expr = ctx.find_node_at_offset::<ast::MatchExpr>()?;
+    let match_arm_list = match_expr.match_arm_list()?;
+    if match_arm_list.arms().count() != 2 {
+        cov_mark::hit!(non_two_arm_match);
+        return None;
+    }
+
+    let mut normal_arm = None;
+    let mut normal_expr = None;
+    let mut wildcard_expr = None;
+    for arm in match_arm_list.arms() {
+        if matches!(arm.pat(), Some(Pat::WildcardPat(_))) && arm.guard().is_none() {
+            wildcard_expr = arm.expr();
+        } else if !matches!(arm.pat(), Some(Pat::WildcardPat(_))) {
+            normal_arm = Some(arm.clone());
+            normal_expr = arm.expr();
+        }
+    }
```

Comment:
This is rather limiting, ideally we want this assist to also work in scenarious like

```rs
match option {
    Some(_) => true,
    None => (),
}
```

Since we only ever expect two patterns here (we could support more but that makes things rather difficult so we shouldn't for the time being until r-a gets a better understanding of patterns) we should just always pick the pattern of the first arm unconditionally. Given arms are tried in top to bottom order we will always pick the correct pattern (+ guard) that way. Then we only need to check if the the body is true or not and adjust accordingly.

That makes this code here even simple, we don't need to bother looking at the second arm at all.

I'll review the rest after this has been addressed, since that will change a lot of things. If there are questions regarding this feel free to ask.

Reply:
Thank you for your review! Updated!